File: Lab 4/D/Alice.java
import java.io.*;
import java.net.*;

public class Alice {
    public static void main(String[] args) {
        try {
            Socket socket = new Socket("localhost", 12345);

            int p = 29; // prime number
            int a = 2; // primitive root

            // Calculating the public key of Alice and sending
            System.out.print("Enter the private key for Alice: ");
            int xa = Integer.parseInt(new BufferedReader(new InputStreamReader(System.in)).readLine());
            int ya = (int) Math.pow(a, xa) % p; // Public key of Alice
            new PrintStream(socket.getOutputStream()).println(ya);

            // Receive public key from Eve
            int yd2 = Integer.parseInt(new BufferedReader(new InputStreamReader(socket.getInputStream())).readLine());
            System.out.println("Received public key for Alice: " + yd2);

            // Calculate the secret key
            int k2 = (int) Math.pow(yd2, xa) % p;
            System.out.println("Secret key decrypted by Alice: " + k2);

            socket.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: Lab 4/Py/alice.py
import socket
port = 12345
client_socket = socket.socket()
client_socket.connect(('localhost', port))

                                            # 1. Calculating the public key of Alice and send 
#As per the algorithm it would be the first step  getting Alices public key and sending to Bob which would be intruded by Darth
xa = int(input("Enter the private key for ALice:"))
ya = pow(2,xa,29) #Public key of Alice
client_socket.send(str(ya).encode())


yd2 = int(client_socket.recv(1024).decode()) #received the public key decided by Darth
print("Received public key for Alice:", yd2) 

                                            #7. Calculate the secret key sent by Bob which does not match with Bob
k2 = pow(yd2, xa, 29)
print(f"Secret key decrypted by ALice: {k2}")
client_socket.close()
